### PR TITLE
python@3.10: make keg-only

### LINKED ISSUE #86520: python@3.10 overrides /usr/local/bin/python3 etc. installed by python@3.9
### `brew gist-logs <formula>` link OR `brew config` AND `brew doctor` output

```shell
`brew config`:


HOMEBREW_VERSION: 3.2.15
ORIGIN: https://github.com/Homebrew/brew
HEAD: 92749f9c5243691f8e576aa1227cb9ebba509cd4
Last commit: 2 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 7f9f35c8b47c0c7be433b608b5ca22adfd409c9b
Core tap last commit: 3 hours ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_CLEANUP_MAX_AGE_DAYS: 1
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.RJVoZZ6LWA/org.xquartz:0
HOMEBREW_EDITOR: /usr/bin/vim
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.3 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit icelake
Clang: 12.0.5 build 1205
Git: 2.33.0 => /usr/local/bin/git
Curl: 7.64.1 => /usr/local/Homebrew/Library/Homebrew/shims/shared/curl
macOS: 11.6-x86_64
CLT: 12.5.0.22.9
Xcode: 13.0
XQuartz: 2.8.1 => /opt/X11
```

`brew doctor`:

```
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: A newer Command Line Tools release is available.
Update them from Software Update in System Preferences or run:
  softwareupdate --all --install --force

If that doesn't show you any updates, run:
  sudo rm -rf /Library/Developer/CommandLineTools
  sudo xcode-select --install

Alternatively, manually download them from:
  https://developer.apple.com/download/all/.
You should download the Command Line Tools for Xcode 13.0.


Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  lua@5.1

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  python@3.10
```
```


### 

- [X] I ran `brew update` and am still able to reproduce my issue.
- [X] I have resolved all warnings from `brew doctor` and that did not fix my problem.

### What were you trying to do (and why)?

`brew install python@3.10`.  As the `python` formula is still an alias to `python@3.9`, I expected that `/usr/local/bin/python3` would continue to point to 3.9 after the upgrade, with 3.10 not symlinked into PATH.

### What happened (include all command output)?

* I installed Python 3.10.  Homebrew complained along the way that `/usr/local/bin/python3` etc. were overwritten during the `brew link` step.
* I found that the `/usr/local/bin/python3` symlink and friends (pip etc.) now pointed to Python 3.10, though individual `python3.9` and `python3.10` programs were both available and ran the respective versions.
* I unlinked 3.10 and relinked 3.9.  `python3` is now 3.9, and there is no `python3.10` in the PATH.
* I suspect that this will repeat should 3.10 receive an update before 3.9.

```
Tue Oct 05 ~/work$ brew install python@3.10
Updating Homebrew...
==> Auto-updated Homebrew!
Updated 2 taps (homebrew/core and homebrew/cask).
==> New Formulae
libnghttp2
==> Updated Formulae
Updated 16 formulae.
==> Updated Casks
Updated 10 casks.

==> Downloading https://ghcr.io/v2/homebrew/core/ca-certificates/manifests/2021-
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/ca-certificates/blobs/sha256:47
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sh
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/openssl/1.1/manifests/1.1.1l_1
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/openssl/1.1/blobs/sha256:64d4d8
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sh
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/readline/manifests/8.1.1
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/readline/blobs/sha256:c596199dc
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sh
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/python/3.10/manifests/3.10.0
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/python/3.10/blobs/sha256:f415f1
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sh
######################################################################## 100.0%
==> Installing dependencies for python@3.10: ca-certificates, openssl@1.1 and readline
==> Installing python@3.10 dependency: ca-certificates
==> Pouring ca-certificates--2021-09-30.all.bottle.1.tar.gz
==> Regenerating CA certificate bundle from keychain, this may take a while...
🍺  /usr/local/Cellar/ca-certificates/2021-09-30: 3 files, 203.5KB
==> Installing python@3.10 dependency: openssl@1.1
==> Pouring openssl@1.1--1.1.1l_1.big_sur.bottle.tar.gz
🍺  /usr/local/Cellar/openssl@1.1/1.1.1l_1: 8,073 files, 18.5MB
==> Installing python@3.10 dependency: readline
==> Pouring readline--8.1.1.big_sur.bottle.tar.gz
🍺  /usr/local/Cellar/readline/8.1.1: 48 files, 1.6MB
==> Installing python@3.10
==> Pouring python@3.10--3.10.0.big_sur.bottle.tar.gz
Warning: These files were overwritten during the `brew link` step:
/usr/local/bin/2to3
/usr/local/bin/idle3
/usr/local/bin/pydoc3
/usr/local/bin/python3
/usr/local/bin/python3-config
/usr/local/share/man/man1/python3.1
/usr/local/lib/pkgconfig/python3-embed.pc
/usr/local/lib/pkgconfig/python3.pc
/usr/local/Frameworks/Python.framework/Headers
/usr/local/Frameworks/Python.framework/Python
/usr/local/Frameworks/Python.framework/Resources
/usr/local/Frameworks/Python.framework/Versions/Current

They have been backed up to: /Users/jwodder/Library/Caches/Homebrew/Backup
==> /usr/local/Cellar/python@3.10/3.10.0/bin/python3 -m ensurepip
==> /usr/local/Cellar/python@3.10/3.10.0/bin/python3 -m pip install -v --no-deps
==> Caveats
Python has been installed as
  /usr/local/bin/python3

Unversioned symlinks `python`, `python-config`, `pip` etc. pointing to
`python3`, `python3-config`, `pip3` etc., respectively, have been installed into
  /usr/local/opt/python@3.10/libexec/bin

You can install Python packages with
  pip3 install <package>
They will install into the site-package directory
  /usr/local/lib/python3.10/site-packages

tkinter is no longer included with this formula, but it is available separately:
  brew install python-tk@3.10

See: https://docs.brew.sh/Homebrew-and-Python
==> Summary
🍺  /usr/local/Cellar/python@3.10/3.10.0: 3,133 files, 55.8MB
==> Upgrading 3 dependents:
imagemagick 7.1.0-8 -> 7.1.0-9, nghttp2 1.45.1 -> 1.45.1_1, unbound 1.13.2 -> 1.13.2_1
==> Downloading https://ghcr.io/v2/homebrew/core/libomp/manifests/13.0.0
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/libomp/blobs/sha256:be00288f6f2
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sh
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/openexr/manifests/3.1.2
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/openexr/blobs/sha256:922846fff4
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sh
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/imagemagick/manifests/7.1.0-9
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/imagemagick/blobs/sha256:476959
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sh
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/libnghttp2/manifests/1.45.1
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/libnghttp2/blobs/sha256:6edccdb
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sh
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/nghttp2/manifests/1.45.1_1
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/nghttp2/blobs/sha256:4c27ecc896
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sh
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/libnghttp2/manifests/1.45.1
Already downloaded: /Users/jwodder/Library/Caches/Homebrew/downloads/a10e8809f2b80ff379c8ac93e0e41e69e5a179b36e2ac19968bcfec8abe11b73--libnghttp2-1.45.1.bottle_manifest.json
==> Downloading https://ghcr.io/v2/homebrew/core/libnghttp2/blobs/sha256:6edccdb
Already downloaded: /Users/jwodder/Library/Caches/Homebrew/downloads/0ddc64f9fd99263fe024b4c46452c0f244578238ff5a2b0321e5742e8e08522e--libnghttp2--1.45.1.big_sur.bottle.tar.gz
==> Downloading https://ghcr.io/v2/homebrew/core/unbound/manifests/1.13.2_1
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/unbound/blobs/sha256:7f411a6ec2
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sh
######################################################################## 100.0%
==> Upgrading imagemagick
  7.1.0-8 -> 7.1.0-9 

==> Installing dependencies for imagemagick: libomp and openexr
==> Installing imagemagick dependency: libomp
==> Pouring libomp--13.0.0.big_sur.bottle.tar.gz
🍺  /usr/local/Cellar/libomp/13.0.0: 9 files, 1.6MB
==> Installing imagemagick dependency: openexr
==> Pouring openexr--3.1.2.big_sur.bottle.tar.gz
🍺  /usr/local/Cellar/openexr/3.1.2: 194 files, 5.9MB
==> Installing imagemagick
==> Pouring imagemagick--7.1.0-9.big_sur.bottle.tar.gz
🍺  /usr/local/Cellar/imagemagick/7.1.0-9: 799 files, 30.3MB
Removing: /usr/local/Cellar/imagemagick/7.1.0-8... (799 files, 30.3MB)
==> Upgrading nghttp2
  1.45.1 -> 1.45.1_1 

==> Installing dependencies for nghttp2: libnghttp2
==> Installing nghttp2 dependency: libnghttp2
==> Pouring libnghttp2--1.45.1.big_sur.bottle.tar.gz
🍺  /usr/local/Cellar/libnghttp2/1.45.1: 13 files, 659.9KB
==> Installing nghttp2
==> Pouring nghttp2--1.45.1_1.big_sur.bottle.tar.gz
🍺  /usr/local/Cellar/nghttp2/1.45.1_1: 17 files, 2.2MB
Removing: /usr/local/Cellar/nghttp2/1.45.1... (23 files, 2.8MB)
Removing: /Users/jwodder/Library/Caches/Homebrew/nghttp2--1.45.1... (972.3KB)
==> Upgrading unbound
  1.13.2 -> 1.13.2_1 

==> Pouring unbound--1.13.2_1.big_sur.bottle.tar.gz
🍺  /usr/local/Cellar/unbound/1.13.2_1: 57 files, 5.7MB
Removing: /usr/local/Cellar/unbound/1.13.2... (57 files, 5.7MB)
==> Checking for dependents of upgraded formulae...
==> No broken dependents found!
==> Caveats
==> python@3.10
Python has been installed as
  /usr/local/bin/python3

Unversioned symlinks `python`, `python-config`, `pip` etc. pointing to
`python3`, `python3-config`, `pip3` etc., respectively, have been installed into
  /usr/local/opt/python@3.10/libexec/bin

You can install Python packages with
  pip3 install <package>
They will install into the site-package directory
  /usr/local/lib/python3.10/site-packages

tkinter is no longer included with this formula, but it is available separately:
  brew install python-tk@3.10

See: https://docs.brew.sh/Homebrew-and-Python
Tue Oct 05 ~/work$ python3 -V
Python 3.10.0
Tue Oct 05 ~/work$ python3.9 -V
Python 3.9.7
Tue Oct 05 ~/work$ python3.10 -V
Python 3.10.0
Tue Oct 05 ~/work$ pip3 -V
pip 21.2.4 from /usr/local/lib/python3.10/site-packages/pip (python 3.10)
Tue Oct 05 ~/work$ brew unlink python@3.10
Unlinking /usr/local/Cellar/python@3.10/3.10.0... 24 symlinks removed.
Tue Oct 05 ~/work$ brew link python@3.9
Warning: Already linked: /usr/local/Cellar/python@3.9/3.9.7
To relink, run:
  brew unlink python@3.9 && brew link python@3.9
Tue Oct 05 ~/work$ python3 -V
-bash: /usr/local/bin/python3: No such file or directory
[127]
Tue Oct 05 ~/work$ brew unlink python@3.9
Unlinking /usr/local/Cellar/python@3.9/3.9.7... 10 symlinks removed.
Tue Oct 05 ~/work$ brew link python@3.9
Linking /usr/local/Cellar/python@3.9/3.9.7... 24 symlinks created.
Tue Oct 05 ~/work$ python3 -V
Python 3.9.7
Tue Oct 05 ~/work$ python3.9 -V
Python 3.9.7
Tue Oct 05 ~/work$ python3.10 -V
-bash: /usr/local/bin/python3.10: No such file or directory
[127]
Tue Oct 05 ~/work$ pip3 --version
pip 21.2.4 from /usr/local/lib/python3.9/site-packages/pip (python 3.9)
```

### What did you expect to happen?

`python@3.10` and `python@3.9` should not have gotten into a slap-fight over the ownership of `/usr/local/bin/python3`.

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
brew install python@3.9
brew install python@3.10
```

### PR DESCRIPTION
**Pull request by branchv**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

Closes #86520

https://github.com/Homebrew/homebrew-core/pull/84626#issuecomment-934033595

cc: @carlocab

### CODE REVIEW

**Comment by Bo98** [comment]
This should involve a revision bump or existing installations won't be made keg only.

**Comment by branchvincent** [comment]
does that imply this can't be a `syntax-only` pr?

**Comment by Bo98** [comment]
Revision bump means new bottles so it can’t be syntax-only, sorry.

It’s important we migrate users or they will face conflicts if they attempt to install ***@***.***

**Comment by Bo98** [comment]
`link_overwrite` can also be removed when made keg-only.

**Comment by Bo98** [comment]
Other adjustments to make:

```diff
```

```diff
   def caveats
     <<~EOS
       Python has been installed as
-        #{HOMEBREW_PREFIX}/bin/python3
+        #{opt_bin}/python3
 
       Unversioned symlinks `python`, `python-config`, `pip` etc. pointing to
       `python3`, `python3-config`, `pip3` etc., respectively, have been install
ed into
         #{opt_libexec}/bin
 
       You can install Python packages with
-        pip3 install <package>
+        #{opt_bin}/pip3 install <package>
```

**Comment by branchvincent** [comment]
happy to fix those, but CI will take several days correct? I was assuming we needed a quick fix, because we currently have `python@3.9` and `python@3.10` both installing as `$(brew --prefix)/python3`.

if that's the case and I'm not misunderstanding, then perhaps we should first make `python@3.10` keg-only?

**Comment by carlocab** [comment]
If we want something quick we could add `conflicts_with` for now to produce a better error message until we can make this `keg-only`.

**Comment by branchvincent** [comment]
seems the last [python migration](https://github.com/Homebrew/homebrew-core/issues/62201) went like this:
1. Add `python@3.x+1` as `keg-only` in a [single PR](https://github.com/Homebrew/homebrew-core/pull/61163)
2. Update one formula per PR  to `depends_on "python@3.x+1`
3. Make `python@3.x+1` the [main python](https://github.com/Homebrew/homebrew-core/pull/63346)

if we're following that again, then we just missed a making `python@3.10` `keg-only`, right? that PR should be quick

**Comment by carlocab** [comment]
> seems the last [python migration](https://github.com/Homebrew/homebrew-core/issues/62201) went like this:
> 
> 1. Add `python@3.x+1` as `keg-only` in a [single PR](https://github.com/Homebrew/homebrew-core/pull/61163)
> 2. Update one formula per PR  to `depends_on "python@3.x+1`
> 3. Make `python@3.x+1` the [main python](https://github.com/Homebrew/homebrew-core/pull/63346)
> 
> if we're following that again, then we just missed making `python@3.10` `keg-only`, right? i think a PR to correct that should be quick

Sounds good!

**Comment by Bo98** [comment]
> but CI will take several days correct?

Well, one day maybe - not several.

Though I'm also fine with the 3.10 keg only approach for now.

**Comment by carlocab** [comment]
We also need to do `python-tk@3.10`.

Also, more changes might be needed here -- IIRC what we do with `ensurepip` here isn't done in the older Python formulae because it installs stuff directly into `HOMEBREW_PREFIX` (rather than into the formula prefix in `HOMEBREW_CELLAR`) so it breaks the formula being keg-only.

At least, I think that's what I found when I tried backporting the changes to `python@3.8`. I may have just done things wrong or am misremembering.

**Comment by Bo98** [comment]
Installing to `/usr/local/lib/python3.9/site-packages` is fine as our keg only formulae still symlink to there (intentionally). If `ensurepip` only touches that then it's fine, if not then it's worth looking to fix that.

**Comment by carlocab** [comment]
I _think_ `ensurepip` installs a `/usr/local/bin/pip` (or `pip3`?).

**Comment by Bo98** [comment]
A quick test in the CI is to ask it to print the contents of `HOMEBREW_PREFIX/bin` at the end of postinstall (and might as well do `python-tk` at the same time since that'll be a couple more runs).

**Comment by branchvincent** [comment]
seems neither [`HOMEBREW_PREFIX/"bin/pip3"`](https://github.com/Homebrew/homebrew-core/runs/3809205647) nor [`HOMEBREW_PREFIX/"bin/pip"`](https://github.com/Homebrew/homebrew-core/pull/86469/checks?check_run_id=3809281060) get created

**Review by carlocab — APPROVED** [review]
Thanks @branchvincent!

File: Formula/python@3.10.rb [hunk 1]
```diff
@@ -23,6 +24,8 @@ class PythonAT310 < Formula
   # build packages later. Xcode-only systems need different flags.
   pour_bottle? only_if: :clt_installed
 
+  keg_only :versioned_formula
+
   depends_on "pkg-config" => :build
```
[the remark this answers is not in the document]
Comment (Bo98) [reply Formula/python@3.10.rb:26]: We shouldn't use OpenSSL 3.0 until it or 1.1 is keg only, or we're just going to reintroduce the problem #86520 is having.

[the remark this answers is not in the document]
Comment (Bo98) [reply Formula/python@3.10.rb:26]: Also that person was probably right in saying we should avoid anything too important while that old macOS build issue is still a thing.

**Pull request closed without merging**